gl/opengl.lisp

```diff
@@ -714,67 +765,124 @@ program PROGRAM as multiple values. 1: Size of attribute. 2: Type of attribute.
         (4 (%gl:uniform-matrix-4fv
             location matrix-count transpose array))))))
 
-(macrolet ((def (n % comp)
-             `(defun ,n (location matrices &optional (transpose t))
-                ,(format nil
- "Upload a matrix or matrices to uniform LOCATION. MATRICES is a single
+(defun program-uniform-matrix (program location dim matrices &optional (transpose t))
+  (with-foreign-matrices (array matrices dim)
+    (let ((matrix-count (length matrices)))
+      (case dim
+        (2 (%gl:program-uniform-matrix-2fv program
+            location matrix-count transpose array))
+        (3 (%gl:program-uniform-matrix-3fv program
+            location matrix-count transpose array))
+        (4 (%gl:program-uniform-matrix-4fv program
+            location matrix-count transpose array))))))
+
+(macrolet ((def (n % comp &optional (direct-p nil))
+             (flet ((args (rest) (append (when direct-p '(program)) rest)))
```

Comment:
I don't like solution with this `flet` myself, but that's the only thing I could think of that doesn't require rewriting the whole macro, which is already pretty messy.